grabbit/src/main/groovy/com/twcable/grabbit/jcr/ProtoPropertyDecorator.groovy

```diff
@@ -0,0 +1,92 @@
+import com.twcable.grabbit.proto.NodeProtos.Property as ProtoProperty
+import com.twcable.grabbit.proto.NodeProtos.Value as ProtoValue
+import groovy.transform.CompileStatic
+import org.apache.jackrabbit.value.ValueFactoryImpl
+
+import javax.annotation.Nonnull
+import javax.jcr.Node as JCRNode
+import javax.jcr.Property
+import javax.jcr.PropertyType
+import javax.jcr.Value
+import javax.jcr.ValueFormatException
+
+import static org.apache.jackrabbit.JcrConstants.JCR_MIXINTYPES
+import static org.apache.jackrabbit.JcrConstants.JCR_PRIMARYTYPE
+
+@CompileStatic
+class ProtoPropertyDecorator {
+
+    @Delegate
+    ProtoProperty innerProtoProperty
+
+    ProtoPropertyDecorator(@Nonnull ProtoProperty protoProperty) {
+        this.innerProtoProperty = protoProperty
+    }
+
+
+    void writeToNode(@Nonnull JCRNode node) {
+        if(primaryType || mixinType) throw new IllegalStateException("Refuse to write jcr:primaryType or jcr:mixinType as normal properties.  These are not allowed")
+        try {
+            if (!innerProtoProperty.hasValues()) {
+                node.setProperty(innerProtoProperty.name, getPropertyValue(), innerProtoProperty.type)
+            }
+            else {
+                Value[] values = getPropertyValues()
+                node.setProperty(innerProtoProperty.name, values, innerProtoProperty.type)
+            }
+        }
+        catch (ValueFormatException ex) {
+            if (ex.message.contains("Multivalued property can not be set to a single value")) {
+                //If this is the exception, that means that a property with the name already exists
+                final Property currentProperty = node.getProperty(innerProtoProperty.name)
+                if (currentProperty.multiple) {
+                    final Value[] values = [getPropertyValue()]
+                    node.setProperty(innerProtoProperty.name, values, innerProtoProperty.type)
+                }
+            }
+            else if (ex.message.contains("Single-valued property can not be set to an array of values")) {
+                node.setProperty(innerProtoProperty.name, getPropertyValues().first(), innerProtoProperty.type)
+            }
+        }
+    }
+
+
+    boolean isPrimaryType() {
+        innerProtoProperty.name == JCR_PRIMARYTYPE
+    }
+
+
+    boolean isMixinType() {
+        innerProtoProperty.name == JCR_MIXINTYPES
+    }
+
+
+    private Value getPropertyValue() throws ValueFormatException {
+        getJCRValueFromProtoValue(innerProtoProperty.getValue())
+    }
+
+
+    private Value[] getPropertyValues() throws ValueFormatException {
+        return innerProtoProperty.values.valueList.collect { ProtoValue protoValue -> getJCRValueFromProtoValue(protoValue) } as Value[]
+    }
+
+
+    private Value getJCRValueFromProtoValue(ProtoValue value) throws ValueFormatException {
+
+        final valueFactory = ValueFactoryImpl.getInstance()
+
+        if (innerProtoProperty.type == PropertyType.BINARY) {
```

Comment:
On the speed argument for `switch`: Java has [a special bytecode for `switch` (actually, two)](http://blog.jamesdbloom.com/JavaCodeToByteCode_PartOne.html#switch), so it does enjoy some advantages...

However, because Groovy lets you do a lot more with a `switch` statement than Java, it writes its byte code a bit differently. I haven't looked into what happens with `@CompileStatic`, which may give effectively identical bytecode for cases where there's no effectively difference between Groovy and Java.

All that said, though, the JIT is very, very smart. Especially for such a common construct. Similar to classic compiler techniques like loop-unrolling, it will convert `if else` sequences to the native equivalents to `switch` statements for you if doing so makes sense in context.

As always, go for the clearest way of communicating intent -- let the compiler worry about micro-efficiencies unless there's concrete evidence it's not able to do so in a specific case.